pkg/genericcomp/genericcomp.go

```diff
@@ -0,0 +1,162 @@
+// Copyright (c) 2020, Oracle and/or its affiliates.
+// Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl.
+
+package genericcomp
+
+import (
+	v1beta1v8o "github.com/verrazzano/verrazzano-crd-generator/pkg/apis/verrazzano/v1beta1"
+	"github.com/verrazzano/verrazzano-operator/pkg/types"
+	"github.com/verrazzano/verrazzano-operator/pkg/util"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// GenericComponentSelectorLabel defines the selector label for generic component resources.
+const GenericComponentSelectorLabel = "verrazzano.name"
+
+// NewDeployment constructs a deployment for a generic component.
+func NewDeployment(generic v1beta1v8o.VerrazzanoGenericComponent, namespace string, labels map[string]string) *appsv1.Deployment {
+	return &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      generic.Name,
+			Namespace: namespace,
+			Labels:    labels,
+		},
+		Spec: appsv1.DeploymentSpec{
+			Replicas: func() *int32 {
+				if generic.Replicas == nil {
+					return util.NewVal(1)
+				}
+				return generic.Replicas
+			}(),
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					GenericComponentSelectorLabel: generic.Name,
+				},
+			},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						GenericComponentSelectorLabel: generic.Name,
+					},
+				},
+				Spec: generic.Deployment,
+			},
+		},
+	}
+}
+
+// NewService constructs a service for a generic component.
+func NewService(generic v1beta1v8o.VerrazzanoGenericComponent, namespace string, labels map[string]string) *corev1.Service {
+	service := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      generic.Name,
+			Namespace: namespace,
+			Labels:    labels,
+		},
+		Spec: corev1.ServiceSpec{
+			Selector: map[string]string{
+				GenericComponentSelectorLabel: generic.Name,
+			},
+			Ports: []corev1.ServicePort{},
+			Type:  corev1.ServiceTypeClusterIP,
+		},
+	}
+
+	// Add all container ports to the k8s service
+	for _, container := range generic.Deployment.Containers {
+		for _, port := range container.Ports {
+			sp := corev1.ServicePort{
+				Name:     port.Name,
+				Protocol: port.Protocol,
+				Port:     port.ContainerPort,
+			}
+			service.Spec.Ports = append(service.Spec.Ports, sp)
+		}
+	}
+
+	// No ports then no need for a service
```

Comment:
Should the logic related to deciding if this a need for a service or not be done first?

Reply:
Fixed